### PR TITLE
Add type annotations

### PR DESCRIPTION
**Pull request by jcerise**
Addresses issue #65

Adds type annotations across the file. 

Items of note:
* I inferred types to the best of my ability in a few areas, any feedback on incorrect types is appreciated
* Bare asterisk usage in function signature triggers mypy to say that the signature is missing an annotation, not quite sure how to properly suppress that.

### CODE REVIEW

File: adafruit_rfm9x.py
Comment (jerryneedell) [inline adafruit_rfm9x.py:873]: Why is assert used here? If packet = None is not an error condition? Won't asset throw an error in that case?
Perhaps I just don't understand.

Comment (tekktrik) [reply adafruit_rfm9x.py:873]: I agree that this doesn't make sense.  In fact, if `packet` is `None` then the code below should fail, particularly `packet[0]`.  By the time it gets to this line, packet may in fact be `None` but I agree with @jerryneedell that in that case the `if` statement would catch it.

It may be because mypy doesn't know that `self._read_u8()` can only return a positive number.  A negative value would conceivably fall through the cracks but that would never happen.

Comment (jcerise) [reply adafruit_rfm9x.py:873]: Right on, as I said, this is just a little trick to get mypy to be quiet about these sorts of things. I'll remove it.

Comment (jerryneedell) [reply adafruit_rfm9x.py:873]: Since I don't know what mypy is, I'll defer to the other librarians on this.  At first I thought it might cause an unexpected error but that does not appear to be the case. We have worked very hard to remove unnecessary code in this library to keep it small enough for the M0 boards. I'd support removing it if it is not necessary.

Comment (tekktrik) [reply adafruit_rfm9x.py:873]: Sweet!  This is on my list to review!

**Review by tekktrik — CHANGES_REQUESTED** [review]
Awesome job with this, what a doozy! :cold_sweat:

Feedback and comments attached, let me know if I can help explain anything!

**Comment by tekktrik** [comment]
Also, the bare asterisk signifies that any arguments following it, must be keyword-only, so it doesn't make sense to type annotate it since it's more of a sentinel character than actual argument.  But I don't know how to argue with (or otherwise suppress) mypy!  Maybe there is a suitable type annotation for it, but if so, we haven't been using it so no worries!

**Comment by jerryneedell** [comment]
I do hope this will get tested on hardware before merging. Normally I would be happy to do it, but I will not have access to my hardware for a free weeks. Good luck!

**Comment by jcerise** [comment]
Thanks so much for the feedback! I addressed everything as best I could. I saw another contributor say this earlier in a different PR, but this is legitimately the best first experience I have had with a project like this. I really appreciate the time and care taken to address this PR.

There's certainly some tribal knowledge that is now much clearer than when I started, and I learned a few things :)

I agree that testing on hardware would be good, I would be happy to jump forward and volunteer, but I started this at Pycon where there was hardware available, and unfortunately, I no longer have access to the hardware to test this. It is just type annotations though (aside from the requirements changes, and the new try/catch at the top) and should not (famous last words? I hope not :D ) affect the actual functionality of this library. However, I would probably feel better waiting on having this merged until it can be properly tested.

**Comment by jerryneedell** [comment]
@jcerise first of all, thank you for jumping in at PyCon to help with this. Don't let the comments of a curmudgeon like me dampen your enthusiasm ;-)
As I noted before, I'll defer to the other CircuitPython Librarians to comment and/or accept the "type" changes without testing. I don't fully understand what they are for or what impact they have on code size or execution.
I jumped in when executable code changes like the assert were added. I would argue that something like that clearly required on hardware testing. 
If there are no changes to the code execution, again, I'll defer to others to make that call. I don't want to hold things up especially when I am unable to provide the testing.
My one concern is that the changes are adding to the code size enough to impact the feather_m0_rfm9x build of CIrcuitPython which includes this library as a "frozen" module. I would suggest that someone verify that it still builds with this PR. Again, I apologize that I am unable to do that myself at this time.

**Comment by tekktrik** [comment]
Big agree - thank you for jumping in to help at PyCon, @jcerise!  Glad it's been an enjoyable experience - it should!

Type annotations are generally safe, but I would feel better building it if it's frozen in - I just never remember which libraries those are, so thank you @jerryneedell for mentioning that!  I'll build it once it's ready to be safe.

I can take care of testing that, @jcerise, but if you're wondering what we mean by "frozen in", here's some information: https://learn.adafruit.com/building-circuitpython/adding-frozen-modules

Basically one build of the CircuitPython firmware has this cooked into it, so extra should be given to memory size (type annotations are ignored by the compiler I believe) and functionality (since this is cooked into firmware for specific boards, we get the benefit of the user not needed to provide this library, but it also means that any bugs will proliferate much more readily 😅)

**Comment by FoamyGuy** [comment]
Echoing what a few have said, thanks so much for working on this @jcerise!

My thoughts on a few things mentioned above: 

Regarding suppressing mypy, I think we don't need to worry about it for now. We don't have the checks put into our CI yet so for the time being it's okay if it's still complaining about some things. When the time comes to add it to our CI checks we'll discuss the way we want to configure it which I assume will suppress any messages that we can't or don't intend to comply with. 

regarding testing, we'll definitely test on hardware. I think I have some of the hardware to test this on. But it's also okay if we wait for a bit for jerry to return. I definitely believe that he's got more experience and hardware for testing more thoroughly than I. 

We can do the frozen build test as well to make sure we won't run into size trouble. I'm not actually sure what impact the types have on final size of the mpy file. I think they may get stripped somewhere in the process but I'm not sure if the net impact is really 0 or maybe something more.

File: adafruit_rfm9x.py
Comment (tekktrik) [inline adafruit_rfm9x.py:669]: Looks like there was a mixup!  This should be `bool`, it's the other arguements after that should be `Optional[int]`.  I've unresolved those so it's easier to keep track.

File: adafruit_rfm9x.py
Comment (tekktrik) [reply adafruit_rfm9x.py:763]: Looks like this got missed, wanna add it in?

**Review by tekktrik — CHANGES_REQUESTED** [review]
Just a few minor changes that I think got missed.  After those, we can try building the firmware :)

File: adafruit_rfm9x.py
Comment (tekktrik) [reply adafruit_rfm9x.py:688]: This is referring to `node`

File: adafruit_rfm9x.py
Comment (tekktrik) [reply adafruit_rfm9x.py:690]: This is referring to **both** `identifier` and `flags`

File: adafruit_rfm9x.py
Comment (jcerise) [reply adafruit_rfm9x.py:669]: Yup, I got mixed up, thanks for pointing out!

**Comment by tekktrik** [comment]
Nah, no worries, thanks for the quick turnaround!

File: adafruit_rfm9x.py
Comment (tekktrik) [inline adafruit_rfm9x.py:260]: Ah just noticing this now.  @FoamyGuy, this didn't seem to affect the build, and `mpy-cross` is able to compile the file just fine.  We've had issues in the past with type annotating attributes though, so I wanted to get your insight.

**Review by tekktrik — COMMENTED** [review]
Small question for @FoamyGuy

File: adafruit_rfm9x.py
Comment (tekktrik) [inline adafruit_rfm9x.py:270]: Ah just noticing this now. @FoamyGuy, this didn't seem to affect the build, and `mpy-cross` is able to compile the file just fine. We've had issues in the past with type annotating attributes though, so I wanted to get your insight.

Comment (tekktrik) [reply adafruit_rfm9x.py:270]: It may just be best to remove it then - no loose ends is better than few!

Other than than, it just seems to be type annotations, so if we can test with the hardware I think we're golden!

**Comment by tekktrik** [comment]
Looks good to me!  I'd approve it at this point, but it looks @FoamyGuy has it from here for testing, so just holding off so it doesn't get merged beforehand.  Thanks @jcerise!

**Review by FoamyGuy — APPROVED** [review]
Looks good to me. 

I tested several of the examples included in the repo successfully with https://www.adafruit.com/product/3231 on PyGamer, EdgeBadge, and Feather RP2040 all on `7.3.0-beta.2`

**Comment by FoamyGuy** [comment]
Thanks again for working on this @jcerise!

**Pull request merged**
